Linkmgrd subscribing State DB route event

zjswhhh opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

This PR is to make linkmgrd subscribe events from ROUTE_TABLE in State DB, and react accordingly:  
- If any of the two default route state appears to be 'na', linkmgrd should switch to standby. 
- If both are 'ok', there will be a mux state probing and what happens next depends on linkmgrd state machine and the probing response.  

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Test case(new/improvement)
- [x] New feature


### Approach
#### What is the motivation for this PR?

To make linkmgrd subscribe state DB route event, and handle the switchovers. 

#### How did you do it?

#### How did you verify/test it?

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

Related PR: https://github.com/Azure/sonic-swss/pull/2009

yxieca commented on src/MuxManager.cpp:
```diff
@@ -234,6 +234,33 @@ void MuxManager::processProbeMuxState(const std::string &portName, const std::st
+    }
+
+    std::string nextState = "na";
+    if (mIpv4DefaultRouteState == "ok" && mIpv6DefaultRouteState == "ok") {
```
@lguohan do we have concern if either IPv4 or IPv6 BGP is not configured? Can we assume we will have them both eventually?

zjswhhh replied:
Updated per discussion in Gemini sync-up meeting.

yxieca commented on src/link_manager/LinkManagerStateMachine.cpp:
```diff
@@ -815,6 +815,29 @@ void LinkManagerStateMachine::handleSwitchActiveRequestEvent()
+{
+    MUXLOGWARNING(boost::format("%s: state db default route state: %s") % mMuxPortConfig.getPortName() % routeState);
+
+    if(mComponentInitState.test(MuxStateComponent)) {
```
add space after if

zjswhhh replied:
Updated accordingly.

yxieca commented on src/link_manager/LinkManagerStateMachine.cpp:
```diff
@@ -815,6 +815,29 @@ void LinkManagerStateMachine::handleSwitchActiveRequestEvent()
+
+    if(mComponentInitState.test(MuxStateComponent)) {
+        if (ms(mCompositeState) == mux_state::MuxState::Label::Active &&
+            routeState == "na") {
```
I suggest putting '{' on a new line, especially when if condition expands multiple lines. new line helps to identify the beginning of the if body.

zjswhhh replied:
Updated accordingly.

yxieca commented on src/MuxManager.cpp:
```diff
@@ -234,6 +234,34 @@ void MuxManager::processProbeMuxState(const std::string &portName, const std::st
+    }
+
+    PortMapIterator portMapIterator = mPortMap.begin();
+    while(portMapIterator != mPortMap.end()) {
```
space after while please.

zjswhhh replied:
Updated accordingly.

prsunny commented on src/link_manager/LinkManagerStateMachine.cpp:
```diff
@@ -815,6 +815,28 @@ void LinkManagerStateMachine::handleSwitchActiveRequestEvent()
+    MUXLOGWARNING(boost::format("%s: state db default route state: %s") % mMuxPortConfig.getPortName() % routeState);
+
+    if (mComponentInitState.test(MuxStateComponent)) {
+        if (ms(mCompositeState) == mux_state::MuxState::Label::Active && routeState == "na") {
```
I'm thinking we should check for `!= Standby`. What if mux is in a transient state and we got this event? What do you think?

zjswhhh replied:
Yes I think this is a good point. I will evaluate the behavior and update accordingly.

prsunny commented on src/link_manager/LinkManagerStateMachine.cpp:
```diff
@@ -815,6 +815,28 @@ void LinkManagerStateMachine::handleSwitchActiveRequestEvent()
+        if (ms(mCompositeState) == mux_state::MuxState::Label::Active && routeState == "na") {
+            CompositeState nextState = mCompositeState;
+            enterLinkProberState(nextState, link_prober::LinkProberState::Wait);
+            switchMuxState(nextState, mux_state::MuxState::Label::Standby);
```
How does it behave if both ToRs have default route as "na" ?

zjswhhh replied:
Both ToRs will switch to standby, the one switches first will be "active" at the end. Do we expect both ToRs to be "standby" in this scenario?

zjswhhh commented:
boost::asio::ip::make_address is unable to parse "::/0" and "0.0.0.0/0", and considering the key will always be these two strings, I updated the code and removed make_address(). Changes have been tested on virtual dual testbeds and worked as expected. 

@prsunny could you review again? Thanks!
